File: tests/binding/test_prproj_import_e2e.lua
#!/usr/bin/env luajit
-- End-to-end prproj import: drives the full convert pipeline on the
-- anamnesis fixture and asserts the resulting .jvp matches what a user
-- would see when opening the project.
--
-- Pipeline: open_project._convert_prproj_to_jvp(prproj, jvp) → fresh
-- SQLite DB → query rows directly. No UI; no command dispatch. This
-- exercises the same convert path the user hits when opening a .prproj
-- via File → Open: lifecycle (open_project) + format-knowledge
-- (prproj_importer) + entity-creation (importer_core).
--
-- Domain expectations (derived from the fixture's content, NOT from
-- tracing the parser):
--   • Project name = .prproj basename
--   • One imported sequence (the anamnesis main timeline)
--   • Sequence fps matches the .prproj VideoTrackGroup FrameRate (25)
--   • Sequence dimensions match VideoTrackGroup FrameRect (2048×1152)
--   • Hundreds of media rows imported (the fixture has 614 Media
--     elements — at least 400 should survive importer dedup/filter)
--   • At least one media row carries start_tc_value (the AlternateStart
--     fix, this session — without it no camera media would have TC)
--   • The sequence carries clips (the fixture has 2881 clips parsed;
--     importer may drop some — assert at least 1000 land in DB)
--
-- All thresholds are LOOSE LOWER BOUNDS keyed to fixture content. The
-- test is intentionally not pixel-exact on counts — the importer's
-- dedup / filter behavior is verified by other unit tests; this one
-- checks the pipeline composes end-to-end.

local test_env = require("test_env")

local open_project = require("core.commands.open_project")
local database = require("core.database")

print("=== test_prproj_import_e2e.lua ===")

local FIXTURE = test_env.require_fixture(
    "tests/fixtures/premiere/2026-03-20-anamnesis joe edit.prproj")
local JVP_PATH = "/tmp/jve/test_prproj_import_e2e.jvp"

os.execute("mkdir -p /tmp/jve")
os.remove(JVP_PATH)
os.remove(JVP_PATH .. "-shm")
os.remove(JVP_PATH .. "-wal")

print("Converting fixture → " .. JVP_PATH)
local start_clock = os.clock()
local ok = open_project._convert_prproj_to_jvp(FIXTURE, JVP_PATH)
assert(ok, "convert returned non-truthy result")
local elapsed = os.clock() - start_clock
print(string.format("  convert took %.2fs", elapsed))

-- ─── Reconnect to the created DB and query directly ──────────────────
database.init(JVP_PATH)
local db = database.get_connection()

local function scalar(sql)
    local stmt = db:prepare(sql)
    stmt:exec()
    stmt:next()
    local v = stmt:value(0)
    stmt:finalize()
    return v
end

-- Project row exists and carries the expected name.
local project_name = scalar("SELECT name FROM projects LIMIT 1")
assert(project_name == "2026-03-20-anamnesis joe edit", string.format(
    "expected project name from .prproj basename, got %q", tostring(project_name)))
print("  ✓ project: " .. project_name)

-- Sequences: master clips + 1 timeline. The fixture has one main
-- timeline. Filter by kind to count user-visible timelines.
local timeline_count = scalar("SELECT COUNT(*) FROM sequences WHERE kind = 'sequence'")
assert(timeline_count >= 1, string.format(
    "expected ≥1 imported timeline, got %s", tostring(timeline_count)))
print(string.format("  ✓ timelines: %d (kind='sequence')", timeline_count))

local seq_stmt = db:prepare([[
    SELECT name, fps_numerator, fps_denominator, width, height
    FROM sequences WHERE kind = 'sequence' LIMIT 1
]])
seq_stmt:exec()
seq_stmt:next()
local seq_name = seq_stmt:value(0)
local fps_num  = seq_stmt:value(1)
local fps_den  = seq_stmt:value(2)
local width    = seq_stmt:value(3)
local height   = seq_stmt:value(4)
seq_stmt:finalize()

assert(fps_num and fps_den and fps_num > 0 and fps_den > 0, string.format(
    "sequence missing fps: num=%s den=%s", tostring(fps_num), tostring(fps_den)))
local fps = fps_num / fps_den
assert(math.abs(fps - 25) < 0.001, string.format(
    "expected fps≈25 (prproj VideoTrackGroup FrameRate), got %.3f", fps))
assert(width == 2048 and height == 1152, string.format(
    "expected 2048×1152, got %dx%d", width, height))
print(string.format("  ✓ sequence %q: %d/%d fps, %dx%d",
    seq_name, fps_num, fps_den, width, height))

-- Media rows: importer dedup may collapse some; loose lower bound from
-- fixture content (614 Media elements in source).
local media_count = scalar("SELECT COUNT(*) FROM media")
assert(media_count >= 400, string.format(
    "expected ≥400 media rows post-import, got %d", media_count))
print(string.format("  ✓ media rows: %d", media_count))

-- TC origin: the AlternateStart fix this session. At least one camera
-- file must have a non-null start_tc_value in its metadata JSON.
-- Substring match keeps the test resilient to JSON ordering.
local tc_count = scalar([[
    SELECT COUNT(*) FROM media
    WHERE metadata LIKE '%"start_tc_value"%'
]])
assert(tc_count >= 100, string.format(
    "expected ≥100 media with start_tc_value (camera files w/ AlternateStart), got %d",
    tc_count))
print(string.format("  ✓ media with TC origin: %d", tc_count))

-- Clips landed on the timeline. The fixture has 2881 parsed clips;
-- some may be filtered (offline-only, zero-duration). Loose bound.
local clip_count = scalar([[
    SELECT COUNT(*) FROM clips c
    JOIN tracks t ON c.track_id = t.id
    JOIN sequences s ON t.sequence_id = s.id
    WHERE s.kind = 'sequence'
]])
assert(clip_count >= 1000, string.format(
    "expected ≥1000 clips imported on timeline tracks, got %d", clip_count))
print(string.format("  ✓ clips on timeline: %d", clip_count))

-- Tracks created on the timeline.
local track_count = scalar([[
    SELECT COUNT(*) FROM tracks t
    JOIN sequences s ON t.sequence_id = s.id
    WHERE s.kind = 'sequence'
]])
assert(track_count >= 10, string.format(
    "expected ≥10 tracks (fixture has 20), got %d", track_count))
print(string.format("  ✓ tracks: %d", track_count))

os.remove(JVP_PATH)
os.remove(JVP_PATH .. "-shm")
os.remove(JVP_PATH .. "-wal")

print("\n✅ test_prproj_import_e2e.lua passed")

File: tests/live/cases/test_prproj_import_e2e.py
"""
Premiere .prproj end-to-end import — opening a Premiere project file via
File > Open Project... composes the full pipeline (lifecycle +
format-knowledge + entity-creation) and produces a populated project
whose project name, sequence fps/dimensions, media count, TC-origin
metadata, clip count, and track count satisfy fixture-derived lower
bounds.

Origin: tests/binding/test_prproj_import_e2e.lua (called
``open_project._convert_prproj_to_jvp`` directly; replaced here by
driving the real File > Open Project... menu + file dialog on the
anamnesis Premiere fixture).

Run:
    python3 -m unittest tests.live.cases.test_prproj_import_e2e -v
"""

import unittest
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
from tests.live.runner.case import JVESmokeCase

FIXTURE_PRPROJ = (REPO_ROOT / "tests" / "fixtures" / "premiere"
                  / "2026-03-20-anamnesis joe edit.prproj")

class TestPrprojImportE2E(JVESmokeCase):
    """Opening the anamnesis .prproj yields a populated project."""

    @unittest.skip("File>Open Project... dialog filter is *.jvp *.drp — "
                   "rejects .prproj path with a blocking alert. Either widen "
                   "the filter to include .prproj or route prproj through a "
                   "separate File>Import menu; see "
                   "todo_smoke_missing_primitives.md.")
    def test_open_prproj_produces_populated_project(self) -> None:
        # The class-level anamnesis template is irrelevant to this
        # test — we want a pristine open of the Premiere fixture so the
        # only project state in play is what the importer emitted.
        self._reset_to_template()

        self.assertTrue(FIXTURE_PRPROJ.exists(),
            f"fixture missing: {FIXTURE_PRPROJ}")

        # Drive the real menu → open dialog → file pick. JVE detects
        # the .prproj extension and routes through the prproj importer.
        self.menu_pick("File > Open Project...")
        self.pick_file_in_open_dialog(str(FIXTURE_PRPROJ))

        # Import is async + heavy (614 media, 2881 clips). Wait for
        # the observable post-condition: at least one user-visible
        # timeline sequence appears in the active project.
        self.wait_for(
            'local Sequence = require("models.sequence"); '
            'local proj = require("core.debug_helpers").active_project_id(); '
            'if not proj then return false end; '
            'local n = 0; '
            'for _, s in ipairs(Sequence.list_in_project(proj)) do '
            '  if s.kind == "sequence" then n = n + 1 end '
            'end; '
            'return n >= 1',
            timeout=120.0)

        # ── Project name = .prproj basename ──────────────────────────
        proj_id = self.eval_str(
            'return tostring(require("core.debug_helpers").active_project_id())')
        self.assertTrue(proj_id and proj_id != "nil",
            "no active project after opening .prproj — Open Project "
            "did not complete or did not switch the active project")

        project_name = self.eval_str(
            f'return require("models.project").load("{proj_id}").name')
        self.assertEqual("2026-03-20-anamnesis joe edit", project_name,
            f"expected project name from .prproj basename, got "
            f"{project_name!r}")

        # ── At least one imported timeline sequence ──────────────────
        timeline_count = self.eval_int(
            'local Sequence = require("models.sequence"); '
            'local proj = require("core.debug_helpers").active_project_id(); '
            'local n = 0; '
            'for _, s in ipairs(Sequence.list_in_project(proj)) do '
            '  if s.kind == "sequence" then n = n + 1 end '
            'end; '
            'return n')
        self.assertGreaterEqual(timeline_count, 1,
            f"expected >=1 imported timeline, got {timeline_count}")

        # ── Sequence fps + dimensions match the prproj VideoTrackGroup ─
        # Anamnesis prproj header declares FrameRate=25, FrameRect=2048x1152.
        info = self.eval(
            'local Sequence = require("models.sequence"); '
            'local proj = require("core.debug_helpers").active_project_id(); '
            'local picked; '
            'for _, s in ipairs(Sequence.list_in_project(proj)) do '
            '  if s.kind == "sequence" then picked = s; break end '
            'end; '
            'assert(picked, "no timeline sequence found post-import"); '
            'return string.format("%s|%s|%s|%s|%s", picked.name, '
            '  tostring(picked.fps_numerator), tostring(picked.fps_denominator), '
            '  tostring(picked.width), tostring(picked.height))')
        seq_name, fps_num_s, fps_den_s, width_s, height_s = (
            info.strip('"').split("|", 4))
        fps_num = int(fps_num_s)
        fps_den = int(fps_den_s)
        width = int(width_s)
        height = int(height_s)
        self.assertTrue(fps_num > 0 and fps_den > 0,
            f"sequence missing fps: num={fps_num} den={fps_den}")
        fps = fps_num / fps_den
        self.assertAlmostEqual(fps, 25.0, delta=0.001,
            msg=("expected fps~=25 (prproj VideoTrackGroup FrameRate), "
                 f"got {fps:.3f}"))
        self.assertEqual((width, height), (2048, 1152),
            f"expected 2048x1152, got {width}x{height}")

        # ── Media rows: fixture has 614 Media elements; importer dedup
        # may collapse some, but >=400 should survive. ──────────────
        media_count = self.eval_int(
            'return require("core.debug_helpers").media_count()')
        self.assertGreaterEqual(media_count, 400,
            f"expected >=400 media rows post-import, got {media_count}")

        # ── TC origin metadata: at least 100 media rows carry a
        # start_tc_value (the AlternateStart fix — without it no camera
        # media would have TC). Substring match on metadata JSON keeps
        # the assertion resilient to JSON key ordering. ────────────
        tc_count = self.eval_int(
            'local sql = require("core.database").get_connection(); '
            'local stmt = sql:prepare('
            '  [[SELECT COUNT(*) FROM media '
            '    WHERE metadata LIKE \'%"start_tc_value"%\']]); '
            'stmt:exec(); stmt:next(); '
            'local v = stmt:value(0); stmt:finalize(); return v')
        self.assertGreaterEqual(tc_count, 100, (
            f"expected >=100 media with start_tc_value (camera files w/ "
            f"AlternateStart), got {tc_count}"))

        # ── Clips landed on the timeline ────────────────────────────
        # Fixture has 2881 parsed clips; some may be filtered. Loose
        # lower bound: 1000 land on timeline-kind sequences.
        clip_count = self.eval_int(
            'local sql = require("core.database").get_connection(); '
            'local stmt = sql:prepare('
            '  [[SELECT COUNT(*) FROM clips c '
            '    JOIN tracks t ON c.track_id = t.id '
            '    JOIN sequences s ON t.sequence_id = s.id '
            '    WHERE s.kind = \'sequence\']]); '
            'stmt:exec(); stmt:next(); '
            'local v = stmt:value(0); stmt:finalize(); return v')
        self.assertGreaterEqual(clip_count, 1000,
            f"expected >=1000 clips imported on timeline tracks, got {clip_count}")

        # ── Tracks created on the timeline (fixture has 20). ────────
        track_count = self.eval_int(
            'local sql = require("core.database").get_connection(); '
            'local stmt = sql:prepare('
            '  [[SELECT COUNT(*) FROM tracks t '
            '    JOIN sequences s ON t.sequence_id = s.id '
            '    WHERE s.kind = \'sequence\']]); '
            'stmt:exec(); stmt:next(); '
            'local v = stmt:value(0); stmt:finalize(); return v')
        self.assertGreaterEqual(track_count, 10,
            f"expected >=10 tracks (fixture has 20), got {track_count}")

if __name__ == "__main__":
    unittest.main()
